Doc improvements and update to latest OCI spec PR

rata opened the pull request:
Hi!

Here are some fixes to the docs when I tried to use the agent, with some common pitfalls that might hit people running in other distros. Hopefully this will save some time to someone else in the future :)

Also, updated the code to use the new OCI spec PR. The main reason for this is that the runc code to use the agent as it is on master is not pushed anywhere now, and we want to use the new code. So I thought it would be simpler just to update it here in master. However, I hit an issue with the client-go we were using (fixed, so a nice improvement!) and updated RecvMsg code to not leak open fds, as I needed to adjust just that part for the new runtime-spec (not a big issue, but it was quite easy to do DoS sending in the socket fds before, until the agent reaches the max open files).

With this PR hopefully we have more clear instructions and can use then runc branches stated :) 

For more info please see each commit msg

rata commented on pkg/agent/agent.go:
```diff
@@ -39,10 +95,6 @@ func receiveNewSeccompFile(resolver registry.ResolverFunc, sockfd int) (*registr
-	seccompFdIndex, ok := containerProcessState.FdIndexes["seccompFd"]
-	if !ok || seccompFdIndex < 0 {
-		return nil, nil, fmt.Errorf("recvfd: didn't receive seccomp fd")
-	}
```
Buuut we still leak fds if we exit there: https://github.com/kinvolk/seccompagent/pull/7/files#diff-18bf746206c8ac217eb16ffb0cf91a6e676e05a7b517dc70aef0260d0871028fL40

alban replied:
Is this comment still valid? I don't understand where the link should point to.

rata replied:
No, I think it is not valid anymore. Sorry I didn't update that :)

alban commented on pkg/agent/agent.go:
```diff
@@ -16,6 +16,10 @@ import (
 func receiveNewSeccompFile(resolver registry.ResolverFunc, sockfd int) (*registry.Registry, *os.File, error) {
 	MaxNameLen := 4096
+
+	// We want to receive an int (a file descriptor).
+	// 4 = sizeof(int) in most architectures
+	// We should probably revisit this when porting
```
Suggested change:

```suggestion
	// File descriptors over SCM_RIGHTS are 'int' according to "man cmsg".
	// The unix golang package assumes that a file descriptor is a int32, see:
	// https://github.com/golang/sys/blob/68d13333faf2/unix/sockcmsg_unix.go#L66-L73
	// On Linux and Windows, `int` is always 32 bits, so it's fine:
	// https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models
```

alban replied:
You could add `// +build linux` at the top of the file, since this will work only on Linux.

rata replied:
Not sure I follow, isn't all of this Linux only? Why add that tag? Or add them to all files, you mean?

alban replied:
Adding the tag would make sure that people don't try to use this SCM_RIGHTS code on other OSes because I am not sure it would work elsewhere.

seccompagent is Linux only, but sometimes people on Windows/macOS like to be able to run unit tests with "go test" on specific packages (even if they can't compile the full project), so it could be useful to add the Linux tag only on package that actually require Linux.

I see that [runc's cmsg.go](https://github.com/opencontainers/runc/blob/249bca0a1316129dcd5bd38b5d75572274181cb5/libcontainer/utils/cmsg.go#L36) uses `unix.CmsgSpace(4)` and has the Linux tag, so I would just do the same tag.

SCM_RIGHTS can actually work on other Unixes:
https://github.com/golang/sys/blob/68d13333faf2/unix/sockcmsg_unix.go#L5-L6
```
//go:build aix || darwin || dragonfly || freebsd || linux || netbsd || openbsd || solaris || zos
// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris zos
```
But I suggest we don't complicate this and just add a Linux tag.

rata replied:
Ohh, haven't thought of people using other OSs :see_no_evil: ! Good idea, will add that!

rata replied:
Added locally, will be in the next push

alban commented on go.mod:
```diff
@@ -17,6 +17,6 @@ require (
 )
 
-replace github.com/opencontainers/runtime-spec => github.com/kinvolk/runtime-spec v1.0.2-0.20201110202115-2755fc508653
+replace github.com/opencontainers/runtime-spec => github.com/kinvolk/runtime-spec v1.0.2-0.20210309175439-58798e75e980
```
You could update this to some commit after https://github.com/opencontainers/runtime-spec/pull/1096 and remove the 'replace' directive, now that it is merged in runtime-spec upstream.

rata replied:
Updated to that now! :)

alban reviewed (changes requested):
* The 4 commits starting with "docs" LGTM.
* I added a couple of comments on the commits "go.mod" and "pkg/agent", LGTM once fixed.
* I have not reviewed the last commit "Use lastest OCI runtime-spec PR" yet.

rata commented:
@alban PTAL. 

I've also added a comment to the last commit, saying that we can potentially leak fds if we return before we have a reference to the fds slice. I left that as a following improvement

alban commented on pkg/agent/agent.go:
```diff
@@ -30,6 +30,56 @@ import (
+// StateFds is assumed to be formated as specs.ContainerProcessState.Fds and
+// recvFds the corresponding list of received fds in the same SCM_RIGHT message.
+func parseStateFds(stateFds []string, recvFds []int) (uintptr, error) {
+	// Lets find the index in stateFds of the seccomp-fd.
+	idx := -1
+	err := false
+
+	for i, name := range stateFds {
+		if name == specs.SeccompFdName && idx == -1 {
+			idx = i
+			continue
+		}
+
+		// We found the seccompFdName two times. Error out!
+		if name == specs.SeccompFdName && idx != -1 {
+			err = true
+		}
+	}
+
+	if idx == -1 || err {
+		return 0, fmt.Errorf("seccomp fd not found or malformed containerProcessState.Fds")
+	}
```
Would this be more concise / readable?

```suggestion
	// Lets find the index in stateFds of the seccomp-fd.
	idx := -1
	idxCount := 0
	
	for i, name := range stateFds {
		if name == specs.SeccompFdName {
			idx = i
			idxCount++
		}
	}
	if idxCount != 1 {
		return 0, errors.New("seccomp fd not found or malformed containerProcessState.Fds")
	}
```

rata replied:
Why not.

But it is buggy, checking for idxCount 1= 1 doesn't fail if `name == specs.SeccompFdName` is never true. Making it instead `if idxCount != 1 || idx == -1` :)

rata commented on pkg/agent/agent.go:
```diff
@@ -30,6 +30,56 @@ import (
+	}
+}
+
+// parseContainerProcessState returns the seccomp-fd and closes the rest of the fds in recvFds.
```
Added locally, thanks!

rata commented:
@alban rebased and addressed all the review comments. PTAL :)

I've also added a new commit: using upstream libseccomp-golang instead of our fork. The changes have been merged upstream already and I was already there changing the go.mod to fix the conflicts in this branch, so went ahead and added it.

rata commented:
Wait, CI is broken. Let me fix that first :)

alban reviewed (approved):
LGTM, thanks!
